src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -25,13 +28,19 @@
     from .base import Requirement
     from .factory import Factory
 
+    BaseCandidate = Union[
+        "AlreadyInstalledCandidate",
+        "EditableCandidate",
+        "LinkCandidate",
+    ]
+
```

Comment:
Is there any advantage at some point in making this an actual base class, rather than just extending this list?

This would be something for a follow-up PR, though, no need to do anything about it now.

Reply:
One difference is that a base class requires defining instance variables in the base class’s `__init__`. Union is much more convinience IMO.

On the other hand though, `AlreadyInstalledCandidate` does share some logic with `_InstallRequirementBackedCandidate` that some refactoring would be useful at some point. I’d say it’s better to wait until we are sure the current abstraction is correct though.

Reply:
Sounds reasonable, and I'm fine with leaving it as a MyPy Union only.